guides/deployment/docker.md

```diff
@@ -0,0 +1,304 @@
+RUN cd ${PHOENIX_SUBDIR}/assets \
+  && yarn install \
+  && yarn deploy \
+  && cd .. \
+  && mix phx.digest
+RUN mix release --env=prod --verbose \
+  && mv _build/prod/rel/${APP_NAME} /opt/release \
+  && mv /opt/release/bin/${APP_NAME} /opt/release/bin/start_server
+FROM alpine:3.7
+# bash is required by distillery
+RUN apk update && apk --no-cache --update add bash openssl-dev
+ENV PORT=4000 MIX_ENV=prod REPLACE_OS_VARS=true
+WORKDIR /opt/app
+EXPOSE ${PORT}
+COPY --from=0 /opt/release .
+CMD ["/opt/app/bin/start_server", "foreground"]
+```
+
+(Make sure to check which version of elixir you want to use)
+
+Here are the steps taken here
+
+- get and compile elixir dependencies
+- get frontend dependencies and build frontend assets
+- make a release
+- move that release to an alpine image for image size efficiency
+
+### Test that your build works
+
+You need to push that image to a registery in order to run it on another machine than yours. At hub.docker.com you can create an account and you get one free private image.
+
+To test your build use:
+
+- `docker build -t my_docker_hub_handle/my_app:1 .`
+
+To push that image to docker hub:
+
+- `docker login` (you will need to enter your username and password)
+- `docker push my_docker_hub_handle/my_app:1`
+
+## Running that image on the server
+
+### First run on local
+
+we need to run our image with a database image. To compose images together, we use docker-compose.
+add a `docker-compose.yml` with the following content to the root of your project.
+
+```
+version: '3.1'
+
+services:
+  web:
+    image: "my_docker_hub_handle/my_app:1"
+    ports:
+      - "80:4000"
+      - "443:443"
+    volumes:
+    - .:/app
+    env_file:
+     - ./docker.env
+    stdin_open: true
+    tty: true
+    links:
+      - db
+
+  db:
+    image: postgres:10-alpine
+    volumes:
+      - "./volumes/postgres:/var/lib/postgresql/data"
+    ports:
+      - "5432:5432"
+    env_file:
+     - ./docker.env
+```
+
+this is where the previously created environement variable file will be used.
+
+The first time you run the postgres container, it will create a database and a user for you if you have the POSTGRES_USER, POSTGRES_PASSWORD and POSTGRES_DB variables set up. That's why it's important to use exactly these environement variable names.
+
+first you need to initialize your db container (that will create the initial database)
+`docker-compose up db`
+once the message that the database is ready appears, you can Ctrl-c to exit the container
+
+Now you can run your setup with
+`docker-compose up`
+
+go visit `localhost` on your browser and verify that your application is running.You can then exit the running containers with Ctrl-c.
+
+### Running your images on the server
+
+Prerequisite: you need to have docker and docker-compose installed on your server. (digital ocean has these one-click droplets for example, that are vps you can setup with one click that already have docker and docker-compose)
```

Comment:
By no means I am a devops, but isn't kubernetes or docker swarm for when you want to deploy with several servers?
could you detail a bit the downsides of running docker-compose? Then perhaps we can just list downsides (if they are not that harmful) and add later a section on how to use Kubernetes or Docker swarm?
There is a good google cloud tutorial on [how to deploy phoenix to kubernetes](https://cloud.google.com/community/tutorials/elixir-phoenix-on-kubernetes-google-container-engine), this is actually where I found the best dockerfile. However, I found that kubernetes adds another layer of complexity.
- You are now deploying to several machines, and that leads to further questions, are they connected in erlang terms, well no.
- Also depending on where you deploy, the configuration will be different (on AWS the database ressource is such and such, with such and such settings). I didn't want to start advocating a particular way, because I thought that it would create more opinionated discussions.

so how about that.
- we leave that section but rename it to `deploying to a single machine`. Add the potential problems that you might run into with docker-compose.
- Then perhaps later somebody can add a section with `deploying to several machines` or `scaling your application` and detail deployement with kubernetes.

I also setup continuous deployement with CircleCI (which have a great free plan btw), but I thought that it was too much promoting a particular service. If no one is against it, I can add my CD setup next.

Reply:
The downsides of "deploying" software with docker-compose on a single-host is that there is no resilient process restarting when processes crash, no way to reason about high availability (due to one server), no integrated way to ensure the disk with your database is in a safe, persistent location, etc.

I am by no means the arbiter of what goes in the docs, but I love the idea of perhaps enumerating the potential problems that you might run into with docker-compose at the end. And perhaps suggest some alternatives (like that GCP article) for further information?

Reply:
I added system restart handling with systemctl on 
0a24397
could you detail a bit what you mean by `no way to reason about high availability`
so how about the following 

Potential downsides of deploying on a single machine:
- There is no integrated way to ensure that the disk with your database is in a safe, persistent location. (to mitigate that, you can backup your server. On Digital Ocean you can enable backup of your droplet when creating it)

if you want to add something else, and you have a bit of time to write the sentence, I am more than happy to add it.

Reply:
May I suggest using a multi-stage Docker file, so that the final docker image is as small as possible? After all if we are using a release, we don't need an erlang/elixir install on the box.

For an example: https://github.com/Financial-Times/docker-elixir-build#a-multi-stage-minimal-image-build